src/Bicep.Core/Emit/TemplateWriter.cs

```diff
@@ -532,8 +653,9 @@ private static ObjectExpression GetTypePropertiesForUnionTypeExpression(UnionTyp
             StringLiteralType or StringType => "string",
             IntegerLiteralType or IntegerType => "int",
             BooleanLiteralType or BooleanType => "bool",
-            ObjectType => "object",
+            ObjectType or DiscriminatedObjectType => "object",
             ArrayType => "array",
+            TypeType typeType => GetNonLiteralTypeName(typeType.Unwrapped),
```

Comment:
Neither `DiscriminatedObjectType` nor `TypeType` is a literal type, so it's best to let those fall through to the catchall branch.